cli/cmd/weight/weight.go

```diff
@@ -78,47 +59,69 @@ func scaleAndAllocate(ctx *clicontext.CLIContext, args []string, pause bool, inc
-			service.Spec.RolloutConfig = &riov1.RolloutConfig{
-				Pause:           pause,
-				Increment:       increment,
-				IntervalSeconds: interval,
-			}
+			service.Spec.RolloutConfig = rolloutConfig
 			return nil
 		})
 		errs = append(errs, err)
-		//reminder -= scale
 	}
+	return mapper.NewErrors(errs...)
+}
 
-	// now allocate any remaining weight across those pre-weighted services
-	//added := 0
-	//for i, rev := range toAllocate {
-	//	resource, err := ctx.ByID(rev.Name)
-	//	if err != nil {
-	//		return err
-	//	}
-	//	weight := 0
-	//	if i == len(toAllocate)-1 {
-	//		weight = reminder - added
-	//	} else {
-	//		weight = int(float64(*rev.Status.ComputedWeight) / float64(total) * float64(reminder))
-	//		added += weight
-	//	}
-	//	err = ctx.UpdateResource(resource, func(obj runtime.Object) error {
-	//		s := obj.(*riov1.Service)
-	//		if s.Spec.Weight == nil {
-	//			s.Spec.Weight = new(int)
-	//		}
-	//		*s.Spec.Weight = weight
-	//		s.Spec.RolloutConfig = &riov1.RolloutConfig{
-	//			Pause:           pause,
-	//			Increment:       increment,
-	//			IntervalSeconds: interval,
-	//		}
-	//		return nil
-	//	})
-	//
-	//	errs = append(errs, err)
-	//}
+func setPercentageWeight(ctx *clicontext.CLIContext, arg string, rolloutConfig *riov1.RolloutConfig) error {
+	serviceName, scale, err := validateArg(arg)
+	if err != nil {
+		return err
+	}
+	resource, err := ctx.ByID(serviceName)
+	if err != nil {
+		return err
+	}
+	if scale == 100 {
+		return promote.PerformPromote(ctx, serviceName, rolloutConfig)
+	}
+	svcs, err := util.ListAppServicesFromServiceName(ctx, serviceName)
+	if err != nil {
+		return err
+	}
+	// first find all weight on other versions of this service
+	otherSvcTotalWeight := 0
+	svc := resource.Object.(*riov1.Service)
+	for _, s := range svcs {
+		if s.Name == svc.Name {
+			continue
+		}
+		if s.Status.ComputedWeight != nil && *s.Status.ComputedWeight > 0 {
+			otherSvcTotalWeight += *s.Status.ComputedWeight
+		}
+	}
+	// now calculate what computed weight should be to hit or percentage target, and set that
+	weight := scale
+	if otherSvcTotalWeight > 0 {
+		weight = int(float64(otherSvcTotalWeight)/(1-(float64(scale)/100))) - otherSvcTotalWeight
+	}
+	return ctx.UpdateResource(resource, func(obj runtime.Object) error {
+		s := obj.(*riov1.Service)
+		if s.Spec.Weight == nil {
+			s.Spec.Weight = new(int)
+		}
+		*s.Spec.Weight = weight
+		s.Spec.RolloutConfig = rolloutConfig
+		return nil
+	})
+}
 
-	return mapper.NewErrors(errs...)
+func validateArg(arg string) (string, int, error) {
+	serviceName, scaleStr := kv.Split(arg, "=")
+	scaleStr = strings.TrimSuffix(scaleStr, "%")
+	if scaleStr == "" {
+		return serviceName, 0, errors.New("weight params must be in the format of SERVICE=WEIGHT, for example: myservice=10% or myservice=20")
+	}
+	scale, err := strconv.Atoi(scaleStr)
+	if err != nil {
+		return serviceName, scale, fmt.Errorf("failed to parse %s: %v", arg, err)
+	}
+	if scale > 100 {
```

Comment:
if service is a number that exceed 100, looks like error is reported?

Reply:
if scale is greater then yes it errors. We don't allow `svc@v2=200` or `svc@v2=200%`. We should try to push people back down to 100 as a total.